### PR TITLE
Add expression evaluator to debugger (REPL)

### LINKED ISSUE #4473: Add an expression evaluator (REPL support with inspector sequent results)
### Describe the project you are working on

Meow... I made an expression evaluator:

https://github.com/godotengine/godot/pull/60134

image

### Describe the problem or limitation you are having in your project

We need an expression evaluator in debugger, it is a very common need 😊

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

The evaluator has sequent results for expressions. It can get locals and members. When you click objects on results, it opens the object on inspector.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Evaluator is using `Expression` class.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Meow

### Is there a reason why this should be core and not an add-on in the asset library?

Meow

### PR DESCRIPTION
**Pull request by KoBeWi**
Supersedes #60134
Supersedes #76736
Closes https://github.com/godotengine/godot-proposals/issues/4473

For now it's just rebased branch, putting it as draft until I clean it up and fix issues.

### CODE REVIEW

File: editor/debugger/script_editor_debugger.cpp [hunk 1]
```diff
@@ -854,8 +857,9 @@ void ScriptEditorDebugger::_notification(int p_what) {
 			error_tree->connect(SceneStringName(item_selected), callable_mp(this, &ScriptEditorDebugger::_error_selected));
 			error_tree->connect("item_activated", callable_mp(this, &ScriptEditorDebugger::_error_activated));
 			breakpoints_tree->connect("item_activated", callable_mp(this, &ScriptEditorDebugger::_breakpoint_tree_clicked));
-			[[fallthrough]];
-		}
+			connect("started", callable_mp(expression_evaluator, &EditorExpressionEvaluator::on_start));
```
Comment (Faless) [inline editor/debugger/script_editor_debugger.cpp:860]: Is removing `fallthrough` correct? It was there to apply the style via `NOTIFICATION_THEME_CHANGED`, so unless something changed and `NOTIFICATION_THEME_CHANGED` is always sent on enter tree we still need it (or we can move the code to a `_udpate_theme` function and than call it both from there and from the theme changed notification)

**Review by Faless — CHANGES_REQUESTED** [review]
Looking good overall, see review comments.

As mentioned in the other PR, this feature should also be disabled in release build (and probably extracted as an EditorDebuggerPlugin), but we can do that in a separate PR.

File: editor/debugger/script_editor_debugger.cpp (hunk 1 above)
Comment (KoBeWi) [reply editor/debugger/script_editor_debugger.cpp:860]: > unless something changed and NOTIFICATION_THEME_CHANGED is always sent on enter tree

Yes, it's been like that for a long time.
See also #89746

**Comment by Faless** [comment]
> This feature is a part of _debugger_. Does it even exist in release builds?

Yes the debugger is not fully disabled in release builds, only parts of it (like live scene editing), it's historically always been like that, but of course this should be re-evaluated

**Review by Faless — APPROVED** [review]
LGTM

**Review by akien-mga — APPROVED** [review]
Looks great!

**Review by AThousandShips — APPROVED** [review]

**Comment by KoBeWi** [comment]
You can only evaluate when the game is breaked (at breakpoint, error etc.). Maybe this should be more obvious, but it can be improved later.

**Comment by rohanrhu** [comment]
Thank you.

**Comment by akien-mga** [comment]
I forgot to say thanks when merging this, I had wrongly closed the tab /o\

Thanks to everyone involved in implementing this feature, @rohanrhu, @rxlecky and @KoBeWi!

**Comment by rxlecky** [comment]
Great stuff, it's amazing to see this feature finally implemented, after I was unable to do so. Thanks @rohanrhu ! ❤️

**Comment by GyuchanLim** [comment]
image

Hello, I've been using this feature on the `Godot 4.4 dev 3` snapshot and I am loving this feature.
And I have a potential newbie question.

Is there a way to see the Object value of `Display.dealer_total` just like how you can with `dealt_card`

**Comment by KoBeWi** [comment]
You should be able to see it in the inspector when you click the object.

**Comment by Musclorman** [comment]
Google translate:
Hello. Why not copy the python interpreter?

French original text:
Bonjour. Pourquoi pas copier l'interpreteur python?


image

**Comment by KoBeWi** [comment]
Because it serves a different purpose. It's a tool for debugging a running project.

If you want a GDScript interpreter, you can use one available online, like https://gd.tumeo.space/

**Comment by ineptNonsense** [comment]
This bug seems to have not been completely fixed 👀.
Adding this comment here instead of opening a new PR since maybe we would like to simply re-open this PR :).

I've seen the following behaviors in `Godot 4.5.beta1` (which supposedly includes this PR):
(1) When evaluating expressions while inside a static function, the evaluator does not evaluate.
(2) Additionally, not only does the evaluator not evaluate, but the debugger itself immediately continues running as if the user pressed "Continue".
 
Here is a test project to check it on: (as mentioned, please run this on `Godot 4.5.beta1`).
[godot-evaluator-static-bugs.zip](https://github.com/user-attachments/files/20842780/godot-evaluator-static-bugs.zip)

Behavior (1) is an important feature I believe should exist, but behavior (2) is especially concerning as I have also observed it when triggering the bug that this PR fixed (happened to me in Godot 4.4.1.stable) meaning that possibly any bug other users will encounter with the evaluator might have unintended side effects.

**Comment by KoBeWi** [comment]
> Adding this comment here instead of opening a new PR since maybe we would like to simply re-open this PR :).

That's what issues are for.

**Comment by ineptNonsense** [comment]
> > Adding this comment here instead of opening a new PR since maybe we would like to simply re-open this PR :).
> 
> That's what issues are for.

Then I shall be opening one pronto :)

**Pull request merged**